// @june-so/analytics-next/dist/cjs/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsNode = void 0;
var tslib_1 = require("tslib");
var analytics_1 = require("./analytics");
var validation_1 = require("./plugins/validation");
var analytics_node_1 = require("./plugins/analytics-node");
var event_queue_1 = require("./core/queue/event-queue");
var priority_queue_1 = require("./lib/priority-queue");
var AnalyticsNode = /** @class */ (function () {
    function AnalyticsNode() {
    }
    AnalyticsNode.load = function (settings) {
        var _a, _b;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var cookieOptions, queue, options, analytics, nodeSettings, ctx;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        cookieOptions = {
                            persist: false,
                        };
                        queue = new event_queue_1.EventQueue(new priority_queue_1.PriorityQueue(3, []));
                        options = { user: cookieOptions, group: cookieOptions };
                        analytics = new analytics_1.Analytics(settings, options, queue);
                        nodeSettings = {
                            writeKey: settings.writeKey,
                            apiHost: (_a = settings.apiHost) !== null && _a !== void 0 ? _a : "api.june.so/sdk",
                            httpScheme: (_b = settings.httpScheme) !== null && _b !== void 0 ? _b : "https",
                            name: 'analytics-node-next',
                            type: 'after',
                            version: 'latest',
                        };
                        return [4 /*yield*/, analytics.register(validation_1.validation, analytics_node_1.analyticsNode(nodeSettings))];
                    case 1:
                        ctx = _c.sent();
                        analytics.emit('initialize', settings, cookieOptions !== null && cookieOptions !== void 0 ? cookieOptions : {});
                        return [2 /*return*/, [analytics, ctx]];
                }
            });
        });
    };
    return AnalyticsNode;
}());
exports.AnalyticsNode = AnalyticsNode;
