// @june-so/analytics-next/dist/cjs/plugins/validation/__tests__/index.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("..");
var context_1 = require("../../../core/context");
var validEvent = {
    type: 'track',
    anonymousId: 'abc',
    event: 'test',
    properties: {},
    traits: {},
};
describe('validation', function () {
    ;
    ['track', 'identify', 'group', 'page', 'alias'].forEach(function (method) {
        describe(method, function () {
            it('validates that the `event` exists', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var val;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            val = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    // @ts-ignore
                                    return [2 /*return*/, __1.validation[method](
                                        // @ts-ignore
                                        new context_1.Context())];
                                });
                            }); };
                            return [4 /*yield*/, expect(val()).rejects.toMatchInlineSnapshot("[Error: Event is missing]")];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); });
            it('validates that `event.event` exists', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var val;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            val = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    // @ts-ignore
                                    return [2 /*return*/, __1.validation[method](new context_1.Context(tslib_1.__assign(tslib_1.__assign({}, validEvent), { event: undefined })))];
                                });
                            }); };
                            if (!(method === 'track')) return [3 /*break*/, 2];
                            return [4 /*yield*/, expect(val()).rejects.toMatchInlineSnapshot("[Error: Event is not a string]")];
                        case 1:
                            _a.sent();
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            }); });
            it('validates that `properties` or `traits` are objects', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var val;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (method === 'alias') {
                                return [2 /*return*/];
                            }
                            val = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    // @ts-ignore
                                    return [2 /*return*/, __1.validation[method](new context_1.Context(tslib_1.__assign(tslib_1.__assign({}, validEvent), { properties: undefined, traits: undefined })))];
                                });
                            }); };
                            return [4 /*yield*/, expect(val()).rejects.toMatchInlineSnapshot("[Error: properties is not an object]")];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); });
            it('validates that it contains an user', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var val;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            val = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    // @ts-ignore
                                    return [2 /*return*/, __1.validation[method](new context_1.Context(tslib_1.__assign(tslib_1.__assign({}, validEvent), { userId: undefined, anonymousId: undefined })))];
                                });
                            }); };
                            return [4 /*yield*/, expect(val()).rejects.toMatchInlineSnapshot("[Error: Missing userId or anonymousId]")];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); });
        });
    });
});
