// @braid/vue-formulate-i18n/src/locales.js
/**
 * ✏️️ After creating a new locale file (please use en.js as the template) add it
 * to the list of exports here:
 */

import ar from './locales/ar'

import ca from './locales/ca'

import cs from './locales/cs'

import da from './locales/da'

import de from './locales/de'

import en from './locales/en'

import es from './locales/es'

import fr from './locales/fr'

import he from './locales/he'

import hu from './locales/hu'

import it from './locales/it'

import ja from './locales/ja'

import ko from './locales/ko'

import nb from './locales/nb'

import nl from './locales/nl'

import lt from './locales/lt'

import pl from './locales/pl'

import pt from './locales/pt'

import ru from './locales/ru'

import sk from './locales/sk'

import sr from './locales/sr'

import sv from './locales/sv'

import th from './locales/th'

import tr from './locales/tr'

import vi from './locales/vi'

import zh from './locales/zh'

/**
 * ✏️ Then add it to the list here:
 */
export { ar, ca, cs, da, de, en, es, fr, he, hu, it, ja, ko, nb, nl, lt, ru, pl, pt, sk, sr, sv, th, tr, vi, zh }
