// @june-so/analytics-next/dist/cjs/core/query-string/__tests__/pickPrefix.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var pickPrefix_1 = require("../pickPrefix");
describe('pickPrefix', function () {
    it('strips a single prefix to creates an object', function () {
        var traits = pickPrefix_1.pickPrefix('ajs_traits_', {
            /* eslint-disable @typescript-eslint/camelcase */
            ajs_traits_address: '12-123 St.',
        });
        var output = {
            address: '12-123 St.',
        };
        expect(traits).toEqual(output);
    });
    it('stripts multiple prefixes to create an object', function () {
        var traits = pickPrefix_1.pickPrefix('ajs_traits_', {
            /* eslint-disable @typescript-eslint/camelcase */
            ajs_traits_address: '12-123 St.',
            ajs_traits_city: 'Vancouver',
            ajs_traits_province: 'BC',
        });
        var output = {
            address: '12-123 St.',
            city: 'Vancouver',
            province: 'BC',
        };
        expect(traits).toEqual(output);
    });
});
