// @june-so/analytics-next/dist/pkg/tester/ajs-perf.js
import { __awaiter, __generator } from "tslib";
import { setup, teardown } from 'jest-dev-server';
// @ts-ignore
import lighthouse from 'lighthouse/lighthouse-core';
// @ts-ignore
import reportGenerator from 'lighthouse/lighthouse-core/report/report-generator';
export function gatherLighthouseMetrics(page) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, lighthouse(page.url(), {
                    port: 9222,
                    logLevel: 'error',
                }).then(function (_a) {
                    var lhr = _a.lhr;
                    return JSON.parse(reportGenerator.generateReport(lhr, 'json'));
                })];
        });
    });
}
export function globalSetup() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, setup({
                        command: "node src/tester/server.js --port=3001",
                        launchTimeout: 5000,
                        port: 3001,
                    })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
export function globalTeardown() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, teardown()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
