// axe-core/lib/commons/color/index.js
/**
 * Namespace for color-related utilities.
 * @namespace commons.color
 * @memberof axe
 */
export { default as centerPointOfRect } from './center-point-of-rect';
export { default as Color } from './color';
export { default as elementHasImage } from './element-has-image';
export { default as elementIsDistinct } from './element-is-distinct';
export { default as filteredRectStack } from './filtered-rect-stack';
export { default as flattenColors } from './flatten-colors';
export { default as getBackgroundColor } from './get-background-color';
export { default as getBackgroundStack } from './get-background-stack';
export { default as getContrast } from './get-contrast';
export { default as getForegroundColor } from './get-foreground-color';
export { default as getOwnBackgroundColor } from './get-own-background-color';
export { default as getRectStack } from './get-rect-stack';
export { default as hasValidContrastRatio } from './has-valid-contrast-ratio';
export { default as incompleteData } from './incomplete-data';
export { default as getTextShadowColors } from './get-text-shadow-colors';
