// @june-so/analytics-next/dist/cjs/tester/ajs-perf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.globalTeardown = exports.globalSetup = exports.gatherLighthouseMetrics = void 0;
var tslib_1 = require("tslib");
var jest_dev_server_1 = require("jest-dev-server");
// @ts-ignore
var lighthouse_core_1 = tslib_1.__importDefault(require("lighthouse/lighthouse-core"));
// @ts-ignore
var report_generator_1 = tslib_1.__importDefault(require("lighthouse/lighthouse-core/report/report-generator"));
function gatherLighthouseMetrics(page) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, lighthouse_core_1.default(page.url(), {
                    port: 9222,
                    logLevel: 'error',
                }).then(function (_a) {
                    var lhr = _a.lhr;
                    return JSON.parse(report_generator_1.default.generateReport(lhr, 'json'));
                })];
        });
    });
}
exports.gatherLighthouseMetrics = gatherLighthouseMetrics;
function globalSetup() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, jest_dev_server_1.setup({
                        command: "node src/tester/server.js --port=3001",
                        launchTimeout: 5000,
                        port: 3001,
                    })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.globalSetup = globalSetup;
function globalTeardown() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, jest_dev_server_1.teardown()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.globalTeardown = globalTeardown;
