// @june-so/analytics-next/dist/cjs/tester/ajs-tester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tester = exports.testerTeardown = exports.getBrowser = void 0;
var tslib_1 = require("tslib");
var micro_memoize_1 = tslib_1.__importDefault(require("micro-memoize"));
var playwright_1 = tslib_1.__importDefault(require("playwright"));
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function makeStub(page) {
    var stub = {
        register: function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, page.evaluate(function (innerArgs) {
                                var _a;
                                // @ts-ignore
                                return (_a = window.analytics).register.apply(_a, innerArgs).then(function (ctx) { return ctx.toJSON(); });
                                // @ts-ignore
                            }, args)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        },
        track: function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var ctx;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, page.evaluate(function (innerArgs) {
                                var _a;
                                // @ts-ignore
                                return (_a = window.analytics).track.apply(_a, innerArgs).then(function (ctx) {
                                    return ctx.toJSON();
                                });
                                // @ts-ignore
                            }, args)];
                        case 1:
                            ctx = _a.sent();
                            return [2 /*return*/, ctx];
                    }
                });
            });
        },
        page: function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var ctx;
                var _this = this;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, page.evaluate(function (innerArgs) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var _a;
                                return tslib_1.__generator(this, function (_b) {
                                    // @ts-ignore
                                    return [2 /*return*/, (_a = window.analytics).page.apply(_a, innerArgs).then(function (ctx) {
                                            return ctx.toJSON();
                                        })
                                        // @ts-ignore
                                    ];
                                });
                            }); }, args)];
                        case 1:
                            ctx = _a.sent();
                            return [2 /*return*/, ctx];
                    }
                });
            });
        },
        identify: function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var ctx;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, page.evaluate(function (innerArgs) {
                                var _a;
                                // @ts-ignore
                                return (_a = window.analytics).identify.apply(_a, innerArgs).then(function (ctx) {
                                    return ctx.toJSON();
                                });
                                // @ts-ignore
                            }, args)];
                        case 1:
                            ctx = _a.sent();
                            return [2 /*return*/, ctx];
                    }
                });
            });
        },
        browserPage: page,
    };
    return stub;
}
exports.getBrowser = micro_memoize_1.default(function (browserType, remoteDebug) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var browser;
    return tslib_1.__generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, playwright_1.default[browserType !== null && browserType !== void 0 ? browserType : 'chromium'].launch({
                    headless: true,
                    devtools: true,
                    args: [
                        '--no-sandbox',
                        '--disable-setuid-sandbox',
                        remoteDebug ? '--remote-debugging-port=9222' : '',
                    ],
                })];
            case 1:
                browser = _a.sent();
                process.on('unhandledRejection', function () {
                    browser && browser.close();
                });
                return [2 /*return*/, browser];
        }
    });
}); });
function testerTeardown() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var browser;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, exports.getBrowser()];
                case 1:
                    browser = _a.sent();
                    return [4 /*yield*/, browser.close()];
                case 2:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.testerTeardown = testerTeardown;
function tester(_writeKey, url, browserType, remoteDebug) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var browser, page;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, exports.getBrowser(browserType, remoteDebug)];
                case 1:
                    browser = _a.sent();
                    return [4 /*yield*/, browser.newPage()];
                case 2:
                    page = _a.sent();
                    return [4 /*yield*/, page.goto(url || "file://" + process.cwd() + "/src/tester/__fixtures__/index.html")];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, page.evaluate("\n    window.AnalyticsNext.AnalyticsBrowser.load({\n      writeKey: '" + _writeKey + "',\n    }).then(loaded => {\n      window.analytics = loaded[0]\n    })\n  ")];
                case 4:
                    _a.sent();
                    return [4 /*yield*/, page.waitForFunction('window.analytics !== undefined')];
                case 5:
                    _a.sent();
                    return [2 /*return*/, makeStub(page)];
            }
        });
    });
}
exports.tester = tester;
