// @sentry/vue/dist/index.bundle.js
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var browser_1 = require("@sentry/browser");
exports.BrowserClient = browser_1.BrowserClient;
exports.defaultIntegrations = browser_1.defaultIntegrations;
exports.forceLoad = browser_1.forceLoad;
exports.lastEventId = browser_1.lastEventId;
exports.onLoad = browser_1.onLoad;
exports.showReportDialog = browser_1.showReportDialog;
exports.flush = browser_1.flush;
exports.close = browser_1.close;
exports.wrap = browser_1.wrap;
exports.addGlobalEventProcessor = browser_1.addGlobalEventProcessor;
exports.addBreadcrumb = browser_1.addBreadcrumb;
exports.captureException = browser_1.captureException;
exports.captureEvent = browser_1.captureEvent;
exports.captureMessage = browser_1.captureMessage;
exports.configureScope = browser_1.configureScope;
exports.getHubFromCarrier = browser_1.getHubFromCarrier;
exports.getCurrentHub = browser_1.getCurrentHub;
exports.Hub = browser_1.Hub;
exports.Scope = browser_1.Scope;
exports.setContext = browser_1.setContext;
exports.setExtra = browser_1.setExtra;
exports.setExtras = browser_1.setExtras;
exports.setTag = browser_1.setTag;
exports.setTags = browser_1.setTags;
exports.setUser = browser_1.setUser;
exports.startTransaction = browser_1.startTransaction;
exports.Transports = browser_1.Transports;
exports.withScope = browser_1.withScope;
exports.SDK_NAME = browser_1.SDK_NAME;
exports.SDK_VERSION = browser_1.SDK_VERSION;
var browser_2 = require("@sentry/browser");
var utils_1 = require("@sentry/utils");
var sdk_1 = require("./sdk");
exports.init = sdk_1.init;
var router_1 = require("./router");
exports.vueRouterInstrumentation = router_1.vueRouterInstrumentation;
var errorhandler_1 = require("./errorhandler");
exports.attachErrorHandler = errorhandler_1.attachErrorHandler;
var tracing_1 = require("./tracing");
exports.createTracingMixins = tracing_1.createTracingMixins;
var windowIntegrations = {};
// This block is needed to add compatibility with the integrations packages when used with a CDN
var _window = utils_1.getGlobalObject();
if (_window.Sentry && _window.Sentry.Integrations) {
    windowIntegrations = _window.Sentry.Integrations;
}
var INTEGRATIONS = tslib_1.__assign(tslib_1.__assign({}, windowIntegrations), browser_2.Integrations);
exports.Integrations = INTEGRATIONS;
