// vue-easytable/packages/ve-dropdown/src/index.jsx
import clickoutside from "../../src/directives/clickoutside.js";
import VeCheckbox from "vue-easytable/packages/ve-checkbox";
import VeRadio from "vue-easytable/packages/ve-radio";
import layerAdjustment from "../../src/mixins/layer-adjustment.js";
import { COMPS_NAME, EMIT_EVENTS } from "./util/constant";
import { clsName } from "./util/index";

export default {
    name: COMPS_NAME.VE_DROPDOWN,
    directives: {
        "click-outside": clickoutside
    },
    mixins: [layerAdjustment],
    props: {
        // 如果是select 组件将特殊处理
        isSelect: {
            type: Boolean,
            default: false
        },
        showOperation: {
            type: Boolean,
            default: false
        },

        width: {
            type: Number,
            default: 90
        },

        // select的最大宽度(超出隐藏)
        maxWidth: {
            type: Number,
            default: 0
        },
        // max height
        maxHeight: {
            type: Number,
            default: 1000
        },

        // 如果为true 会包含 checkbox
        isMultiple: {
            type: Boolean,
            default: false
        },

        // 用户传入v-model 的值 [{value/label/selected}]
        value: {
            type: [Array],
            default: null
        },

        // 文本居中方式 left|center|right
        textAlign: {
            type: String,
            default: "left"
        },

        // 是否支持输入input
        isInput: {
            type: Boolean,
            default: false
        },
        // confirm filter text
        confirmFilterText: {
            type: String,
            default: ""
        },
        // confirm filter text
        resetFilterText: {
            type: String,
            default: ""
        },
        // hide by single selection item click
        hideByItemClick: {
            type: Boolean,
            default: false
        },
        // is show radio when single selection
        showRadio: {
            type: Boolean,
            default: false
        },
        // 当 isControlled=true ,visible 生效
        visible: {
            type: Boolean,
            default: false
        },
        // is controlled
        isControlled: {
            type: Boolean,
            default: false
        },
        // is custom content
        isCustomContent: {
            type: Boolean,
            default: false
        },
        // instance between dropdown items and trigger element
        instance: {
            type: Number,
            default: 12
        }
    },
    data() {
        return {
            internalVisible: false,
            internalOptions: [],
            inputValue: ""
        };
    },
    computed: {
        // is dropdown visible
        isDropdownVisible() {
            return this.isControlled ? this.visible : this.internalVisible;
        },

        // 获取最大宽度(不设置则是无穷大)
        getMaxWidth() {
            var result = Infinity,
                maxWidth = this.maxWidth,
                width = this.width;

            if (maxWidth && maxWidth > 0 && maxWidth > width) {
                result = maxWidth;
            }

            return result;
        },
        // selected labels
        selectedLabels() {
            return this.internalOptions
                .filter(x => x.selected)
                .map(x => {
                    if (x.selected) {
                        return x.label;
                    }
                });
        },
        // operation buttons class
        operationFilterClass() {
            let result = null;
            result = {
                [clsName("filter-disable")]: this.selectedLabels.length === 0
            };

            return result;
        },
        // dropdown items class
        dropdownItemsClass() {
            return {
                [clsName("dd")]: true,
                [clsName("dd-show")]: this.isDropdownVisible
            };
        }
    },
    watch: {
        value: function(val) {
            this.init();
        }
    },
    methods: {
        // 初始化
        init() {
            this.internalOptions = Object.assign([], this.value);

            if (this.isInput) {
                this.setInputValue();
            }
        },

        // operation filter confirm
        confirm() {
            // 使用户传入的v-model 生效
            this.$emit("input", this.internalOptions);
            this.$emit(EMIT_EVENTS.FILTER_CONFIRM, this.internalOptions);
            this.hideDropDown();
        },

        // operation filter reset
        reset() {
            if (this.internalOptions.some(x => x.selected)) {
                this.internalOptions.map(x => {
                    if (x.selected) {
                        x.selected = false;
                    }
                    return x;
                });

                // 使用户传入的v-model 生效
                this.$emit("input", this.internalOptions);

                this.$emit(EMIT_EVENTS.FILTER_RESET, this.internalOptions);
            }

            this.hideDropDown();
        },

        hideDropDown() {
            this.$emit(EMIT_EVENTS.VISIBLE_CHANGE, false);

            if (!this.isControlled) {
                setTimeout(() => {
                    this.internalVisible = false;
                }, 150);
            }
        },

        showDropDown() {
            if (!this.isControlled) {
                this.internalVisible = true;
            }
            this.$emit(EMIT_EVENTS.VISIBLE_CHANGE, true);
        },

        // 设置文本框的值
        setInputValue() {
            var result, labels;

            labels = this.selectedLabels;
            if (Array.isArray(labels) && labels.length > 0) {
                result = labels.join();
            }

            this.inputValue = result;
        },

        // toggle dropdown
        toggleDropdown() {
            if (this.isDropdownVisible) {
                this.hideDropDown();
            } else {
                this.showDropDown();

                this.$nextTick(x => {
                    this.dropDownClick();
                });
            }
        },

        // single select option click
        singleSelectOptionClick(e, item) {
            this.internalOptions = this.internalOptions.map(x => {
                if (item.label === x.label) {
                    x.selected = true;
                } else {
                    x.selected = false;
                }
                return x;
            });

            if (this.hideByItemClick) {
                this.toggleDropdown();
            }

            if (this.isInput) {
                this.setInputValue();
            }

            // 使用户传入的v-model 生效
            this.$emit("input", this.internalOptions);

            this.$emit(EMIT_EVENTS.ITEM_SELECT_CHANGE, this.internalOptions);
        },

        // 获取样式名称
        getTextAlignClass() {
            return clsName(`items-li-a-${this.textAlign}`);
        },

        // dropdown blur
        dropdownBlur() {
            this.hideDropDown();
        },

        // 下拉点击显示
        dropDownClick() {
            var dtEle = this.$el.querySelector(".ve-dropdown-dt"),
                ddItem = this.$el.querySelector(".ve-dropdown-items");
            this.layerAdjustmentOnce(ddItem, dtEle, this.instance);
        },

        // 确定下拉框的位置
        dropdownAdjust() {
            this.dropDownClick();
            var dtEle = this.$el.querySelector(".ve-dropdown-dt"),
                ddItem = this.$el.querySelector(".ve-dropdown-items");
            this.layerAdjustmentBind(ddItem, dtEle, this.instance);
        },

        // checbox 受控属性管理
        checkedChangeControl(item, isChecked) {
            this.internalOptions = this.internalOptions.map((i, index) => {
                if (i.label === item.label) {
                    i.selected = isChecked;
                }
                return i;
            });

            this.$emit(EMIT_EVENTS.ITEM_SELECT_CHANGE, this.internalOptions);
        }
    },

    created() {
        this.init();
    },
    mounted() {
        this.dropdownAdjust();
    },

    render() {
        const {
            isMultiple,
            getTextAlignClass,
            internalOptions,
            checkboxGroupChange,
            isSelect,
            width,
            maxHeight,
            toggleDropdown,
            getMaxWidth,
            reset,
            singleSelectOptionClick,
            isDropdownVisible,
            showOperation,
            clickOutside,
            isCustomContent
        } = this;

        let content = "";

        if (isMultiple) {
            content = internalOptions.map((item, index) => {
                const checkboxProps = {
                    key: item.label,
                    props: {
                        isControlled: true,
                        label: item.label,
                        showLine: item.showLine,
                        isSelected: item.selected
                    },
                    on: {
                        "on-checked-change": isChecked =>
                            this.checkedChangeControl(item, isChecked)
                    }
                };

                return (
                    <li
                        key={index}
                        class={[
                            clsName("items-multiple"),
                            clsName("items-li"),
                            getTextAlignClass()
                        ]}
                    >
                        <VeCheckbox {...checkboxProps} />
                    </li>
                );
            });
        } else {
            content = internalOptions.map((item, index) => {
                const radioProps = {
                    props: {
                        isControlled: true,
                        isSelected: item.selected
                    },
                    on: {
                        "on-radio-change": () => {}
                    }
                };

                return (
                    <li
                        key={index}
                        class={[
                            clsName("items-li"),
                            item.selected ? "active" : ""
                        ]}
                        on-click={e => singleSelectOptionClick(e, item)}
                    >
                        <a
                            class={[clsName("items-li-a"), getTextAlignClass()]}
                            href="javascript:void(0);"
                        >
                            {this.showRadio ? (
                                <VeRadio {...radioProps}>{item.label}</VeRadio>
                            ) : (
                                item.label
                            )}
                        </a>
                    </li>
                );
            });
        }

        const dropdownProps = {
            class: ["ve-dropdown"],
            directives: [
                {
                    name: "click-outside",
                    value: this.dropdownBlur
                }
            ]
        };

        return (
            <dl {...dropdownProps}>
                <dt class="ve-dropdown-dt">
                    <a
                        class={[isSelect ? clsName("dt-selected") : ""]}
                        style={{ width: width + "px" }}
                        on-click={toggleDropdown}
                    >
                        {this.$slots.default}
                    </a>
                </dt>
                <dd class={this.dropdownItemsClass}>
                    <ul
                        class={clsName("items")}
                        style={{
                            "min-width": width + "px",
                            "max-width": getMaxWidth + "px"
                        }}
                    >
                        {/* custome content */}
                        {isCustomContent && this.$slots["custom-content"]}
                        {/* not custom content */}
                        {!isCustomContent && (
                            <div>
                                <div
                                    style={{
                                        "max-height": maxHeight + "px"
                                    }}
                                    class={clsName("items-warpper")}
                                >
                                    {content}
                                </div>
                                {showOperation && (
                                    <li class={clsName("operation")}>
                                        <a
                                            class={[
                                                clsName("operation-item"),
                                                this.operationFilterClass
                                            ]}
                                            href="javascript:void(0)"
                                            on-click={reset}
                                        >
                                            {this.resetFilterText}
                                        </a>
                                        <a
                                            class={clsName("operation-item")}
                                            href="javascript:void(0)"
                                            on-click={this.confirm}
                                        >
                                            {this.confirmFilterText}
                                        </a>
                                    </li>
                                )}
                            </div>
                        )}
                    </ul>
                </dd>
            </dl>
        );
    }
};
