// @june-so/analytics-next/dist/cjs/core/queue/__tests__/extension-flushing.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var lodash_1 = require("lodash");
var priority_queue_1 = require("../../../lib/priority-queue");
var context_1 = require("../../context");
var event_queue_1 = require("../event-queue");
var fruitBasket = new context_1.Context({
    type: 'track',
    event: 'Fruit Basket',
    properties: {
        banana: '🍌',
        apple: '🍎',
        grape: '🍇',
    },
});
var testPlugin = {
    name: 'test',
    type: 'before',
    version: '0.1.0',
    load: function () { return Promise.resolve(); },
    isLoaded: function () { return true; },
};
var ajs = {};
describe('Registration', function () {
    test('can register plugins', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, load, plugin, ctx;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    load = jest.fn();
                    plugin = {
                        name: 'test',
                        type: 'before',
                        version: '0.1.0',
                        load: load,
                        isLoaded: function () { return true; },
                    };
                    ctx = context_1.Context.system();
                    return [4 /*yield*/, eq.register(ctx, plugin, ajs)];
                case 1:
                    _a.sent();
                    expect(load).toHaveBeenCalledWith(ctx, ajs);
                    return [2 /*return*/];
            }
        });
    }); });
    test('fails if plugin cant be loaded', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, plugin, ctx;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    plugin = {
                        name: 'test',
                        type: 'before',
                        version: '0.1.0',
                        load: function () { return Promise.reject(new Error('👻')); },
                        isLoaded: function () { return false; },
                    };
                    ctx = context_1.Context.system();
                    return [4 /*yield*/, expect(eq.register(ctx, plugin, ajs)).rejects.toThrowErrorMatchingInlineSnapshot("\"\uD83D\uDC7B\"")];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    test('allows for destinations to fail registration', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, plugin, ctx;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    plugin = {
                        name: 'test',
                        type: 'destination',
                        version: '0.1.0',
                        load: function () { return Promise.reject(new Error('👻')); },
                        isLoaded: function () { return false; },
                    };
                    ctx = context_1.Context.system();
                    return [4 /*yield*/, eq.register(ctx, plugin, ajs)];
                case 1:
                    _a.sent();
                    expect(ctx.logs()[0].level).toEqual('warn');
                    expect(ctx.logs()[0].message).toEqual('Failed to load destination');
                    return [2 /*return*/];
            }
        });
    }); });
});
describe('Plugin flushing', function () {
    test('ensures `before` plugins are run', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, queue, flushed, failedFlush, messages;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    queue = new priority_queue_1.PriorityQueue(1, []);
                    eq.queue = queue;
                    return [4 /*yield*/, eq.register(context_1.Context.system(), tslib_1.__assign(tslib_1.__assign({}, testPlugin), { type: 'before' }), ajs)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, eq.dispatch(fruitBasket)];
                case 2:
                    flushed = _a.sent();
                    expect(flushed.logs().map(function (l) { return l.message; })).toContain('Delivered');
                    return [4 /*yield*/, eq.register(context_1.Context.system(), tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Faulty before', type: 'before', track: function () {
                                throw new Error('aaay');
                            } }), ajs)];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, eq
                            .dispatch(new context_1.Context({
                            type: 'track',
                        }))
                            .catch(function (ctx) { return ctx; })];
                case 4:
                    failedFlush = _a.sent();
                    messages = failedFlush.logs().map(function (l) { return l.message; });
                    expect(messages).not.toContain('Delivered');
                    return [2 /*return*/];
            }
        });
    }); });
    test('atempts `enrichment` plugins', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, flushed, messages;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    return [4 /*yield*/, eq.register(context_1.Context.system(), tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Faulty enrichment', type: 'enrichment', track: function () {
                                throw new Error('aaay');
                            } }), ajs)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, eq.dispatch(new context_1.Context({
                            type: 'track',
                        }))];
                case 2:
                    flushed = _a.sent();
                    messages = flushed.logs().map(function (l) { return l.message; });
                    expect(messages).toContain('Delivered');
                    return [2 /*return*/];
            }
        });
    }); });
    test('attempts `destination` plugins', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, amplitude, fullstory, flushed, messages;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    amplitude = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Amplitude', type: 'destination', track: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                throw new Error('Boom!');
                            });
                        }); } });
                    fullstory = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'FullStory', type: 'destination' });
                    return [4 /*yield*/, eq.register(context_1.Context.system(), amplitude, ajs)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, eq.register(context_1.Context.system(), fullstory, ajs)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, eq.dispatch(new context_1.Context({
                            type: 'track',
                        }))];
                case 3:
                    flushed = _a.sent();
                    messages = flushed
                        .logs()
                        .map(function (l) { return ({ message: l.message, extras: l.extras }); });
                    expect(messages).toMatchInlineSnapshot("\n      Array [\n        Object {\n          \"extras\": undefined,\n          \"message\": \"Dispatching\",\n        },\n        Object {\n          \"extras\": Object {\n            \"plugin\": \"Amplitude\",\n          },\n          \"message\": \"plugin\",\n        },\n        Object {\n          \"extras\": Object {\n            \"plugin\": \"FullStory\",\n          },\n          \"message\": \"plugin\",\n        },\n        Object {\n          \"extras\": Object {\n            \"error\": [Error: Boom!],\n            \"plugin\": \"Amplitude\",\n          },\n          \"message\": \"plugin Error\",\n        },\n        Object {\n          \"extras\": Object {\n            \"type\": \"track\",\n          },\n          \"message\": \"Delivered\",\n        },\n      ]\n    ");
                    return [2 /*return*/];
            }
        });
    }); });
    test('attempts `after` plugins', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, afterFailed, after, flushed, messages;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    afterFailed = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'after-failed', type: 'after', track: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                throw new Error('Boom!');
                            });
                        }); } });
                    after = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'after', type: 'after' });
                    return [4 /*yield*/, eq.register(context_1.Context.system(), afterFailed, ajs)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, eq.register(context_1.Context.system(), after, ajs)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, eq.dispatch(new context_1.Context({
                            type: 'track',
                        }))];
                case 3:
                    flushed = _a.sent();
                    messages = flushed
                        .logs()
                        .map(function (l) { return ({ message: l.message, extras: l.extras }); });
                    expect(messages).toMatchInlineSnapshot("\n      Array [\n        Object {\n          \"extras\": undefined,\n          \"message\": \"Dispatching\",\n        },\n        Object {\n          \"extras\": Object {\n            \"plugin\": \"after-failed\",\n          },\n          \"message\": \"plugin\",\n        },\n        Object {\n          \"extras\": Object {\n            \"plugin\": \"after\",\n          },\n          \"message\": \"plugin\",\n        },\n        Object {\n          \"extras\": Object {\n            \"error\": [Error: Boom!],\n            \"plugin\": \"after-failed\",\n          },\n          \"message\": \"plugin Error\",\n        },\n        Object {\n          \"extras\": Object {\n            \"type\": \"track\",\n          },\n          \"message\": \"Delivered\",\n        },\n      ]\n    ");
                    return [2 /*return*/];
            }
        });
    }); });
    test('runs `enrichment` and `before` inline', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, flushed;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    return [4 /*yield*/, eq.register(context_1.Context.system(), tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Kiwi', type: 'enrichment', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    ctx.updateEvent('properties.kiwi', '🥝');
                                    return [2 /*return*/, ctx];
                                });
                            }); } }), ajs)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, eq.register(context_1.Context.system(), tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Watermelon', type: 'enrichment', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    ctx.updateEvent('properties.watermelon', '🍉');
                                    return [2 /*return*/, ctx];
                                });
                            }); } }), ajs)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, eq.register(context_1.Context.system(), tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Before', type: 'before', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    ctx.stats.increment('before');
                                    return [2 /*return*/, ctx];
                                });
                            }); } }), ajs)];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, eq.dispatch(new context_1.Context({
                            type: 'track',
                        }))];
                case 4:
                    flushed = _a.sent();
                    expect(flushed.event.properties).toEqual({
                        watermelon: '🍉',
                        kiwi: '🥝',
                    });
                    expect(flushed.stats.metrics.map(function (m) { return m.metric; })).toContain('before');
                    return [2 /*return*/];
            }
        });
    }); });
    test('respects execution order', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var eq, trace, before, enrichment, enrichmentTwo, destination, plugins, _i, plugins_1, xt;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    eq = new event_queue_1.EventQueue();
                    trace = 0;
                    before = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Before', type: 'before', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                trace++;
                                expect(trace).toBe(1);
                                return [2 /*return*/, ctx];
                            });
                        }); } });
                    enrichment = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Enrichment', type: 'enrichment', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                trace++;
                                expect(trace === 2 || trace === 3).toBe(true);
                                return [2 /*return*/, ctx];
                            });
                        }); } });
                    enrichmentTwo = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Enrichment 2', type: 'enrichment', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                trace++;
                                expect(trace === 2 || trace === 3).toBe(true);
                                return [2 /*return*/, ctx];
                            });
                        }); } });
                    destination = tslib_1.__assign(tslib_1.__assign({}, testPlugin), { name: 'Destination', type: 'destination', track: function (ctx) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                trace++;
                                expect(trace).toBe(4);
                                return [2 /*return*/, ctx];
                            });
                        }); } });
                    plugins = lodash_1.shuffle([before, enrichment, enrichmentTwo, destination]);
                    _i = 0, plugins_1 = plugins;
                    _a.label = 1;
                case 1:
                    if (!(_i < plugins_1.length)) return [3 /*break*/, 4];
                    xt = plugins_1[_i];
                    return [4 /*yield*/, eq.register(context_1.Context.system(), xt, ajs)];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    _i++;
                    return [3 /*break*/, 1];
                case 4: return [4 /*yield*/, eq.dispatch(new context_1.Context({
                        type: 'track',
                    }))];
                case 5:
                    _a.sent();
                    expect(trace).toBe(4);
                    return [2 /*return*/];
            }
        });
    }); });
});
