// @june-so/analytics-next/dist/cjs/plugins/routing-middleware/__tests__/index.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/camelcase */
var __1 = require("..");
var to_facade_1 = require("../../../lib/to-facade");
describe('tsub middleware', function () {
    var rules = [
        {
            matchers: [
                {
                    ir: '["=","event",{"value":"Item Impression"}]',
                    type: 'fql',
                },
            ],
            scope: 'destinations',
            target_type: 'workspace::project::destination',
            transformers: [[{ type: 'drop' }]],
            destinationName: 'Google Tag Manager',
        },
    ];
    var middleware = __1.tsubMiddleware(rules);
    it('can apply tsub operations', function () {
        var payload;
        var next = jest.fn().mockImplementation(function (transformed) {
            payload = transformed === null || transformed === void 0 ? void 0 : transformed.obj;
        });
        middleware({
            integration: 'Google Tag Manager',
            next: next,
            payload: to_facade_1.toFacade({
                type: 'track',
                event: 'Item Impression',
            }),
        });
        expect(next).toHaveBeenCalled();
        expect(payload).toMatchInlineSnapshot("null");
    });
    it("won't match different destinations", function () {
        var payload;
        var next = jest.fn().mockImplementation(function (transformed) {
            payload = transformed === null || transformed === void 0 ? void 0 : transformed.obj;
        });
        middleware({
            integration: 'Google Analytics',
            next: next,
            payload: to_facade_1.toFacade({
                type: 'track',
                event: 'Item Impression',
            }),
        });
        expect(next).toHaveBeenCalled();
        expect(payload).toMatchInlineSnapshot("\n      Object {\n        \"event\": \"Item Impression\",\n        \"type\": \"track\",\n      }\n    ");
    });
});
