// axe-core/lib/commons/aria/index.js
/**
 * Namespace for aria-related utilities.
 * @namespace commons.aria
 * @memberof axe
 */
export { default as allowedAttr } from './allowed-attr';
export { default as arialabelText } from './arialabel-text';
export { default as arialabelledbyText } from './arialabelledby-text';
export { default as getAccessibleRefs } from './get-accessible-refs';
export { default as getElementUnallowedRoles } from './get-element-unallowed-roles';
export { default as getExplicitRole } from './get-explicit-role';
export { default as getOwnedVirtual } from './get-owned-virtual';
export { default as getRoleType } from './get-role-type';
export { default as getRole } from './get-role';
export { default as getRolesByType } from './get-roles-by-type';
export { default as getRolesWithNameFromContents } from './get-roles-with-name-from-contents';
export { default as implicitNodes } from './implicit-nodes';
export { default as implicitRole } from './implicit-role';
export { default as isAccessibleRef } from './is-accessible-ref';
export { default as isAriaRoleAllowedOnElement } from './is-aria-role-allowed-on-element';
export { default as isUnsupportedRole } from './is-unsupported-role';
export { default as isValidRole } from './is-valid-role';
export { default as labelVirtual } from './label-virtual';
export { default as label } from './label';
export { default as lookupTable } from './lookup-table';
export { default as namedFromContents } from './named-from-contents';
export { default as requiredAttr } from './required-attr';
export { default as requiredContext } from './required-context';
export { default as requiredOwned } from './required-owned';
export { default as validateAttrValue } from './validate-attr-value';
export { default as validateAttr } from './validate-attr';
