// mpd-parser/test/manifests/708-captions.js
export const parsedManifest = {
  allowCache: true,
  discontinuityStarts: [],
  duration: 6,
  endList: true,
  mediaGroups: {
    'AUDIO': {},
    'CLOSED-CAPTIONS': {
      cc: {
        // eng: {
        //   autoselect: false,
        //   default: false,
        //   instreamId: '1',
        //   language: 'eng',
        //   aspectRatio: 1,
        //   easyReader: 0,
        //   '3D': 0
        // },
        // TODO only this one ends up being represented and not both
        eng: {
          'autoselect': false,
          'default': false,
          'instreamId': 'SERVICE2',
          'language': 'eng',
          'aspectRatio': 1,
          'easyReader': 1,
          '3D': 0
        }
      }
    },
    'SUBTITLES': {},
    'VIDEO': {}
  },
  playlists: [
    {
      attributes: {
        'AUDIO': 'audio',
        'BANDWIDTH': 449000,
        'CODECS': 'avc1.420015',
        'NAME': '482',
        'PROGRAM-ID': 1,
        'RESOLUTION': {
          height: 270,
          width: 482
        },
        'SUBTITLES': 'subs'
      },
      endList: true,
      resolvedUri: '',
      targetDuration: 6,
      mediaSequence: 0,
      timelineStarts: [{ start: 0, timeline: 0 }],
      discontinuitySequence: 0,
      discontinuityStarts: [],
      segments: [
        {
          duration: 6,
          timeline: 0,
          number: 0,
          presentationTime: 0,
          map: {
            uri: '',
            resolvedUri: 'https://www.example.com/1080p.ts'
          },
          resolvedUri: 'https://www.example.com/1080p.ts',
          uri: 'https://www.example.com/1080p.ts'
        }
      ],
      timeline: 0,
      uri: ''
    }
  ],
  segments: [],
  timelineStarts: [{ start: 0, timeline: 0 }],
  uri: ''
};
