// @june-so/analytics-next/dist/cjs/lib/__tests__/group-by.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var group_by_1 = require("../group-by");
describe('groupBy', function () {
    it('groups a collection by key', function () {
        var collection = [
            {
                id: 'go',
                type: 'server',
            },
            {
                id: 'ruby',
                type: 'server',
            },
            {
                id: 'js',
                type: 'browser',
            },
            {
                id: 'react',
                type: 'app',
            },
        ];
        expect(group_by_1.groupBy(collection, 'type')).toEqual({
            app: [
                {
                    id: 'react',
                    type: 'app',
                },
            ],
            browser: [
                {
                    id: 'js',
                    type: 'browser',
                },
            ],
            server: [
                {
                    id: 'go',
                    type: 'server',
                },
                {
                    id: 'ruby',
                    type: 'server',
                },
            ],
        });
    });
    it('accepts a function', function () {
        var collection = [
            {
                id: 1,
                type: 'server',
            },
            {
                id: 2,
                type: 'server',
            },
            {
                id: 3,
                type: 'browser',
            },
            {
                id: 4,
                type: 'app',
            },
        ];
        var oddEven = group_by_1.groupBy(collection, function (c) {
            return c.id % 2 === 0 ? 'even' : 'odd';
        });
        expect(oddEven).toEqual({
            even: [
                {
                    id: 2,
                    type: 'server',
                },
                {
                    id: 4,
                    type: 'app',
                },
            ],
            odd: [
                {
                    id: 1,
                    type: 'server',
                },
                {
                    id: 3,
                    type: 'browser',
                },
            ],
        });
    });
});
