// vue-easytable/packages/ve-table/src/body/body-td.jsx
import BodyCheckboxContent from "./body-checkbox-content";
import BodyRadioContent from "./body-radio-content";
import ExpandTrIcon from "./expand-tr-icon";
import { getFixedTotalWidthByColumnKey, clsName, getRowKey } from "../util";
import {
    getValByUnit,
    isEmptyValue,
    isNumber,
    isBoolean
} from "../../../src/utils/index.js";

import {
    COMPS_NAME,
    COLUMN_TYPES,
    EXPAND_TRIGGER_TYPES,
    EMIT_EVENTS
} from "../util/constant";
import emitter from "../../../src/mixins/emitter";

export default {
    name: COMPS_NAME.VE_TABLE_BODY_TD,
    mixins: [emitter],
    props: {
        rowData: {
            type: Object,
            required: true
        },
        column: {
            type: Object,
            required: true
        },
        rowIndex: {
            type: Number,
            required: true
        },
        colgroups: {
            type: Array,
            required: true
        },

        rowKeyFieldName: {
            type: String,
            default: null
        },
        /*
        expand
        */
        // expand row option
        expandOption: {
            type: Object,
            default: function() {
                return null;
            }
        },
        // is expand row
        isExpandRow: {
            type: Boolean,
            required: true
        },
        // expanded row keys
        expandedRowkeys: {
            type: Array,
            default: function() {
                return [];
            }
        },

        /*
        checkbox
        */
        // checkbox option
        checkboxOption: {
            type: Object,
            default: function() {
                return null;
            }
        },
        internalCheckboxSelectedRowKeys: {
            type: Array,
            default: function() {
                return null;
            }
        },

        /*
        radio
        */
        radioOption: {
            type: Object,
            default: function() {
                return null;
            }
        },
        internalRadioSelectedRowKey: {
            type: [String, Number],
            default: null
        },
        // cell style option
        cellStyleOption: {
            type: Object,
            default: function() {
                return null;
            }
        },
        // event custom option
        eventCustomOption: {
            type: Object,
            default: function() {
                return null;
            }
        },
        // cell selection key data
        cellSelectionKeyData: {
            type: Object,
            default: function() {
                return null;
            }
        },
        // cell span option
        cellSpanOption: {
            type: Object,
            default: function() {
                return null;
            }
        },
        // edit opttion
        editOption: {
            type: Object,
            default: function() {
                return null;
            }
        }
    },
    computed: {
        // is last left fixed column
        isLastLeftFixedColumn() {
            let result = false;

            const { colgroups, column } = this;

            const { fixed, field } = column;

            if (fixed === "left") {
                const { field } = column;
                const leftFixedColumns = colgroups.filter(
                    x => x.fixed === "left"
                );
                const index = leftFixedColumns.findIndex(
                    x => x.field === field
                );

                if (index === leftFixedColumns.length - 1) {
                    result = true;
                }
            }
            return result;
        },
        // is first right fixed column
        isfirstRightFixedColumn() {
            let result = false;

            const { colgroups, column } = this;

            const { fixed, field } = column;

            if (fixed === "right") {
                const { field } = column;
                const leftFixedColumns = colgroups.filter(
                    x => x.fixed === "right"
                );

                if (leftFixedColumns[0].field === field) {
                    result = true;
                }
            }
            return result;
        },

        /*
         * @bodyTdStyle
         * @desc body td style
         */
        bodyTdStyle() {
            const { key, align, fixed } = this.column;

            let result = {};

            const { colgroups, column } = this;

            // text align
            result["text-align"] = align || "center";

            // fixed left total width or right total width
            if (fixed) {
                let totalWidth = 0;
                // column index
                const columnIndex = colgroups.findIndex(x => x.key === key);
                if (
                    (fixed === "left" && columnIndex > 0) ||
                    (fixed === "right" && columnIndex < colgroups.length - 1)
                ) {
                    totalWidth = getFixedTotalWidthByColumnKey(
                        colgroups,
                        key,
                        fixed
                    );

                    totalWidth = getValByUnit(totalWidth);
                }

                result["left"] = fixed === "left" ? totalWidth : "";
                result["right"] = fixed === "right" ? totalWidth : "";
            }

            return result;
        },

        /*
         * @bodyTdClass
         * @desc body td class
         */
        bodyTdClass() {
            const { fixed } = this.column;

            let result = {
                [clsName("body-td")]: true
            };

            const {
                cellStyleOption,
                rowData,
                column,
                rowIndex,
                cellSelectionKeyData,
                rowKeyFieldName
            } = this;

            // column fixed
            if (fixed) {
                result[clsName("fixed-left")] = fixed === "left";
                result[clsName("fixed-right")] = fixed === "right";
                result[
                    clsName("last-left-fixed-column")
                ] = this.isLastLeftFixedColumn;
                result[
                    clsName("first-right-fixed-column")
                ] = this.isfirstRightFixedColumn;
            }

            // cell style option
            if (
                cellStyleOption &&
                typeof cellStyleOption.bodyCellClass === "function"
            ) {
                const customClass = cellStyleOption.bodyCellClass({
                    row: rowData,
                    column,
                    rowIndex
                });
                if (customClass) {
                    result[customClass] = true;
                }
            }

            // cell selection option
            if (cellSelectionKeyData) {
                const { rowKey, columnKey } = cellSelectionKeyData;
                if (
                    rowData[rowKeyFieldName] === rowKey &&
                    column["key"] === columnKey
                ) {
                    result[clsName("cell-selection")] = true;
                }
            }

            return result;
        }
    },
    methods: {
        // get ellipsis content style
        getEllipsisContentStyle() {
            let result = {};

            const { ellipsis } = this.column;

            if (ellipsis) {
                const { lineClamp } = ellipsis;

                let _lineClamp = isNumber(lineClamp) ? lineClamp : 1;
                result["-webkit-line-clamp"] = _lineClamp;
            }

            return result;
        },

        // get render content
        getRenderContent(h) {
            let content = null;

            const { column, rowData, rowIndex } = this;

            // has render function
            if (typeof column.renderBodyCell === "function") {
                const renderResult = column.renderBodyCell(
                    {
                        row: rowData,
                        column,
                        rowIndex
                    },
                    h
                );

                content = renderResult;
            } else {
                content = rowData[column.field];
            }

            // ellipisis
            if (column.ellipsis) {
                const { lineClamp, showTitle } = column.ellipsis;

                // default true
                const isShowTitle = isBoolean(showTitle) ? showTitle : true;

                content = (
                    <span
                        title={isShowTitle ? content : ""}
                        style={this.getEllipsisContentStyle()}
                        class={clsName("body-td-span-ellipsis")}
                    >
                        {content}
                    </span>
                );
            }

            // edit 暂时不开放
            /* if (editOption) {
                const { activeRowkey } = editOption;

                if (
                    !isEmptyValue(activeRowkey) &&
                    activeRowkey == getRowKey(rowData, rowKeyFieldName)
                ) {
                    content = (
                        <input
                            class={clsName("body-td-edit-input")}
                            value={content}
                        />
                    );
                }
            } */

            return content;
        },

        // get chcekbox content
        getCheckboxContent() {
            if (this.column.type === COLUMN_TYPES.CHECKBOX) {
                // checkbox content props
                const checkboxProps = {
                    props: {
                        column: this.column,
                        checkboxOption: this.checkboxOption,
                        rowKey: this.rowData[this.rowKeyFieldName],
                        internalCheckboxSelectedRowKeys: this
                            .internalCheckboxSelectedRowKeys
                    }
                };

                return <BodyCheckboxContent {...checkboxProps} />;
            }
            return null;
        },

        // get radio content
        getRadioContent() {
            if (this.column.type === COLUMN_TYPES.RADIO) {
                // radio props
                const radioProps = {
                    props: {
                        column: this.column,
                        radioOption: this.radioOption,
                        rowKey: this.rowData[this.rowKeyFieldName],
                        internalRadioSelectedRowKey: this
                            .internalRadioSelectedRowKey
                    }
                };

                return <BodyRadioContent {...radioProps} />;
            }
            return null;
        },

        // get cell span
        getCellSpan() {
            const { cellSpanOption, rowData, column, rowIndex } = this;
            let rowspan = 1;
            let colspan = 1;

            if (cellSpanOption) {
                const { bodyCellSpan } = cellSpanOption;

                if (typeof bodyCellSpan === "function") {
                    const result = bodyCellSpan({
                        row: rowData,
                        column,
                        rowIndex
                    });

                    if (typeof result === "object") {
                        rowspan = result.rowspan;
                        colspan = result.colspan;
                    }
                }
            }

            return { rowspan, colspan };
        },
        // cell click
        cellClick(e, fn) {
            fn && fn(e);

            const { column, expandOption, rowData } = this;

            this.dispatch(COMPS_NAME.VE_TABLE_BODY, EMIT_EVENTS.BODY_TD_CLICK, {
                rowData,
                column
            });

            if (column.type !== COLUMN_TYPES.EXPAND) {
                return false;
            }

            if (expandOption) {
                const eventTargetName = e.target.nodeName;

                const trigger = expandOption.trigger;

                // expand row by click icon
                if (!trigger || trigger === EXPAND_TRIGGER_TYPES.ICON) {
                    if (eventTargetName !== "TD") {
                        e.stopPropagation();
                        this.$emit(EMIT_EVENTS.EXPAND_ROW_CHANGE);
                    }
                }
                // expand row by click cell(td)
                else if (trigger === EXPAND_TRIGGER_TYPES.CELL) {
                    e.stopPropagation();
                    this.$emit(EMIT_EVENTS.EXPAND_ROW_CHANGE);
                }
            }
        },
        // dblclick
        cellDblclick(e, fn) {
            fn && fn(e);
        },
        // contextmenu
        cellContextmenu(e, fn) {
            fn && fn(e);
        },
        // mouseenter
        cellMouseenter(e, fn) {
            fn && fn(e);
        },
        // mouseleave
        cellMouseleave(e, fn) {
            fn && fn(e);
        }
    },
    render(h) {
        const {
            column,
            cellClick,
            rowData,
            isExpandRow,
            expandOption,
            expandedRowkeys,
            rowKeyFieldName,
            eventCustomOption,
            rowIndex
        } = this;

        // expand icon props
        const expandIconProps = {
            props: {
                rowData,
                column,
                expandOption,
                expandedRowkeys,
                rowKeyFieldName,
                cellClick
            }
        };

        const { rowspan, colspan } = this.getCellSpan();
        if (!rowspan || !colspan) {
            return null;
        }

        // custom on cell event
        let customEvents = {};
        if (eventCustomOption) {
            const { bodyCellEvents } = eventCustomOption;

            customEvents = bodyCellEvents
                ? bodyCellEvents({ row: rowData, column, rowIndex })
                : {};
        }

        const {
            click,
            dblclick,
            contextmenu,
            mouseenter,
            mouseleave
        } = customEvents;

        const events = {
            click: e => {
                this.cellClick(e, click);
            },
            dblclick: e => {
                this.cellDblclick(e, dblclick);
            },
            contextmenu: e => {
                this.cellContextmenu(e, contextmenu);
            },
            mouseenter: e => {
                this.cellMouseenter(e, mouseenter);
            },
            mouseleave: e => {
                this.cellMouseleave(e, mouseleave);
            }
        };

        // td props
        const tdProps = {
            class: this.bodyTdClass,
            style: this.bodyTdStyle,
            attrs: {
                rowspan,
                colspan
            },
            on: events
        };

        return (
            <td {...tdProps}>
                {/* expadn tr icon */}
                {isExpandRow && <ExpandTrIcon {...expandIconProps} />}
                {/* checkbox content */}
                {this.getCheckboxContent()}
                {/* radio content */}
                {this.getRadioContent()}
                {/* other cell content */}
                {this.getRenderContent(h)}
            </td>
        );
    }
};
