// react-syntax-highlighter/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Light", {
  enumerable: true,
  get: function get() {
    return _light["default"];
  }
});
Object.defineProperty(exports, "LightAsync", {
  enumerable: true,
  get: function get() {
    return _lightAsync["default"];
  }
});
Object.defineProperty(exports, "Prism", {
  enumerable: true,
  get: function get() {
    return _prism["default"];
  }
});
Object.defineProperty(exports, "PrismAsync", {
  enumerable: true,
  get: function get() {
    return _prismAsync["default"];
  }
});
Object.defineProperty(exports, "PrismAsyncLight", {
  enumerable: true,
  get: function get() {
    return _prismAsyncLight["default"];
  }
});
Object.defineProperty(exports, "PrismLight", {
  enumerable: true,
  get: function get() {
    return _prismLight["default"];
  }
});
Object.defineProperty(exports, "createElement", {
  enumerable: true,
  get: function get() {
    return _createElement["default"];
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _defaultHighlight["default"];
  }
});

var _defaultHighlight = _interopRequireDefault(require("./default-highlight"));

var _lightAsync = _interopRequireDefault(require("./light-async"));

var _light = _interopRequireDefault(require("./light"));

var _prismAsyncLight = _interopRequireDefault(require("./prism-async-light"));

var _prismAsync = _interopRequireDefault(require("./prism-async"));

var _prismLight = _interopRequireDefault(require("./prism-light"));

var _prism = _interopRequireDefault(require("./prism"));

var _createElement = _interopRequireDefault(require("./create-element"));
